Scheduler API cleanup

andyross opened the pull request:
Getting started looking at the scheduler.  Step one is some work to reduce the external API footprint in the rest of the kernel, which was a little cluttered.  In theory all of this is just cleanup with at most minimal change to generated code.  No logic changes.

Commit:
xtensa, kernel/sched: Move next switch_handle selection to the scheduler

The xtensa asm2 layer had a function to select the next switch handle
to return into following an exception.  There is no arch-specific code
there, it's just scheduler logic.  Move it to the scheduler where it
belongs.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Cleanup, unify _add_thread_to_ready_q() and _ready_thread()

The scheduler exposed two APIs to do the same thing:
_add_thread_to_ready_q() was a low level primitive that in most cases
was wrapped by _ready_thread(), which also (1) checks that the thread
_is_ready() or exits, (2) flags the thread as "started" to handle the
case of a thread running for the first time out of a waitq timeout,
and (3) signals a logger event.

As it turns out, all existing usage was already checking case #1.
Case #2 can be better handled in the timeout resume path instead of on
every call.  And case #3 was probably wrong to have been skipping
anyway (there were paths that could make a thread runnable without
logging).

Now _add_thread_to_ready_q() is an internal scheduler API, as it
probably always should have been.

This also moves some asserts from the inline _ready_thread() wrapper
to the underlying true function for code size reasons, otherwise the
extra use of the inline added by this patch blows past code size
limits on Quark D2000.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/queue: Clean up scheduler API usage

This was the only spot where the scheduler-internal
_peek_first_pending_thread() API was used.  Given that this kind of
thing is inherently racy (it may not be pending as long as you expect
if a timeout expires, etc...), it would be nice to retire it.

And as it happens all the queue code was using it for was to detect
the case of a non-empty wait_q over which it was looping, which is
trivial to do without API support.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Clean up docs for _pend_thread(), limit scope

The scheduler has a kernel-internal _pend_thread() utility which
sounds like a function which will add an arbitrary thread to a wait_q.
This is essentially unsupportable in SMP, where that thread might
actually be executing on a different CPU.

Thankfully we never used it like that.  The only spots outside the
scheduler that use the API are in pipes and mailbox, which both just
want to pend a DUMMY thread to track the timeout but will never try to
pend a true foreign thread.

Clarify the comment and add an assertion to make sure this promise
isn't broken in the future.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

lpereira commented on kernel/queue.c:
```diff
@@ -148,26 +148,21 @@ void k_queue_append_list(struct k_queue *queue, void *head, void *tail)
-	struct k_thread *first_thread, *thread;
+	struct k_thread *thread;
 
-	first_thread = _peek_first_pending_thread(&queue->wait_q);
```
Maybe get rid of this function now that nothing else calls it?

andyross replied:
That will come.  There are actually a *lot* of dead functions in ksched.h.  I think the way this is evolving is that I'll end up writing a new header and adding back only what is needed.

Commit:
kernel_event_logger: Ignore events before subsystem init

Thread and interrupt events may well arrive before the subsystem
initialization call has been made.  Just swallow such events.

In particular, an incoming change to the way _ready_thread works
causes the main and idle thread initialization to throw thread ready
events (which isn't wrong!), which the current setup can't handle.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Had to add one more patch because the change that added the event logger hook to all thread ready operations also added it to the initialization of the main and idle threads which happen before the logging system is initialized.

lpereira commented on kernel/include/ksched.h:
```diff
@@ -541,4 +526,10 @@ static inline int _is_thread_user(void)
 }
 #endif /* CONFIG_USERSPACE */
+
+/**
```
`#if defined(CONFIG_USE_SWITCH)`?

lpereira commented on subsys/logging/kernel_event_logger.c:
```diff
@@ -47,6 +49,8 @@ static int _sys_k_event_logger_init(struct device *arg)
 	sys_event_logger_init(&sys_k_event_logger, _sys_k_event_logger_buffer,
 		CONFIG_KERNEL_EVENT_LOGGER_BUFFER_SIZE);
 
+	initialized = 1;
```
Good catch.

lpereira reviewed (approved):
Patch LGTM; it doesn't look like it'll introduce any regressions.
